Feature/more prescriptive expectations

roblim opened the pull request:
Add prescriptive renderers for following expectations:

expect_column_values_to_be_null
expect_column_values_to_be_of_type
expect_column_values_to_be_in_set
expect_column_values_to_not_be_in_set
expect_column_values_to_be_increasing
expect_column_values_to_be_decreasing
expect_column_value_lengths_to_equal
expect_column_values_to_match_regex
expect_column_values_to_match_regex_list
expect_column_values_to_not_match_regex_list
expect_column_values_to_match_strftime_format
expect_column_values_to_be_dateutil_parseable
expect_column_values_to_be_json_parseable
expect_column_values_to_match_json_schema
expect_column_distinct_values_to_contain_set
expect_column_distinct_values_to_equal_set
expect_column_mean_to_be_between
expect_column_median_to_be_between
expect_column_stdev_to_be_between
expect_column_most_common_value_to_be_in_set 
expect_column_max_to_be_between
expect_column_min_to_be_between
expect_column_sum_to_be_between
expect_column_kl_divergence_to_be_less_than

roblim committed:
Add support for ‘mostly’ kwarg; add renderers

New: expect_column_values_to_be_null, expect_column_values_to_be_of_type

roblim committed:
Add support for ‘parse_strings_as_datetimes’

roblim committed:
Add renderer for expect_column_values_to_be_in_set

roblim committed:
Add renderer for expect_column_values_to_not_be_in_set

roblim committed:
Add renderer for expect_column_values_to_be_increasing

roblim committed:
Add renderer for expect_column_values_to_be_decreasing

roblim committed:
Move method; add “values” to template_str

roblim committed:
Add renderer for expect_column_kl_divergence_to_be_less_than

roblim committed:
Add renderer for expect_column_values_to_match_json_schema

roblim committed:
Add renderer for expect_column_most_common_value_to_be_in_set

roblim committed:
Add renderer for expect_column_sum_to_be_between

roblim committed:
Fix conditional

- if both min_value and max_value missing, correct template_str was being overwritten because subsequent logic not enclosed in else block

roblim committed:
Add renderer for expect_column_min_to_be_between

roblim committed:
Add ‘distinct’

roblim committed:
Fix typo

roblim committed:
Add renderer for expect_column_value_lengths_to_equal

roblim committed:
Add renderer for expect_column_values_to_match_regex

roblim committed:
Move expect_column_values_to_not_match_regex method

roblim committed:
Add renderer for expect_column_values_to_match_regex_list

roblim committed:
Add renderer for expect_column_values_to_not_match_regex_list

roblim committed:
Add renderer for expect_column_values_to_not_match_regex_list

roblim committed:
Add renderer for expect_column_values_to_be_dateutil_parseable

roblim committed:
Add renderer for expect_column_values_to_be_json_parseable

roblim committed:
Add renderer for expect_column_distinct_values_to_contain_set

roblim committed:
Add renderer for expect_column_distinct_values_to_equal_set

roblim committed:
Add renderer for expect_column_mean_to_be_between

roblim committed:
Nix extraneous code block

roblim committed:
Merge branch 'develop' into feature/more-prescriptive-expectations

roblim committed:
Add sparklines to requirements.txt

roblim committed:
Add sparklines

abegong commented on great_expectations/render/renderer/content_block/bullet_list_content_block.py:
```diff
@@ -143,12 +106,22 @@ def expect_column_unique_value_count_to_be_between(cls, expectation, styling=Non
-            template_str = "must have more than $min_value unique values."
         else:
-            template_str = "must have between $min_value and $max_value unique values."
+            if params.get("mostly"):
```
This is a tiny issue, but I'm confused why we're using `if params.get("mostly"):` here and `if params["min_value"] is None:` everywhere else.

Is the behavior the same?

Since we're using `substitute_none_for_missing`, is it even possible for `mostly` to be missing?

abegong replied:
@roblim Reviewing this more closely, it looks like you've replaced my `if 'mostly' in params` with `if params.get("mostly"):`

This is an improvement, but the problem is: my original logic itself was incorrect.

Because we're calling `substitute_none_for_missing` at the top of each method, we *know* that `mostly` will be present.

Instead of checking whether it exists in the dict, we should always be checking `if mostly == None`. (Also, this means that we ALWAYS have to be disciplined about using `substitute_none_for_missing` and including ALL possible parameters for each method.)

roblim replied:
K. Changed to "if params["mostly"] is not None". Regarding your first comment about refactoring `PrescriptiveBulletListContentBlockRenderer`, that refactoring has been done but is in a different branch because it's for prescriptive rendering of EVRS.

abegong reviewed (commented):
Did I misunderstand yesterday? I thought you were refactoring the methods in `PrescriptiveBulletListContentBlockRenderer` into a new class. It's fine to leave it this way---just want to make sure I'm not missing anything.

roblim committed:
Change params.get(“mostly”) condition

- to: if params[“mostly”] is not None:

roblim committed:
Restore expectation renderers

expect_column_max_to_be_between
expect_column_stdev_to_be_between
expect_column_median_to_be_between

roblim committed:
Merge branch 'develop' into feature/more-prescriptive-expectations
